### PR TITLE
[rl] Add aot_fx_trace + CooR precompile to the RL PolicyTrainer

### PR DESCRIPTION
**Pull request by bobrenjc93**
## What

Adds an **opt-in** `aot_fx_trace` + compile-on-one-rank (CooR) **precompile** path to the TitanRL `PolicyTrainer`, mirroring `experiments/graph_trainer`. When enabled, the trainer builds a **SimpleFSDP** (traceable) model and captures `forward + loss + backward` as a single FX graph (regional-Inductor FlexAttention + CUDA graph), reusing it across steps; setting `precompile_artifact_dir` loads a serialized artifact instead of tracing.

The default path is **unchanged** (eager FSDP2), and the generator never receives `aot_fx_trace`.

## Why

The RL trainer currently runs an eager `model() -> loss -> loss.backward()`. graph_trainer already proves that capturing the whole step as one graph unlocks regional-Inductor, CUDA graphs, comm/compute overlap, and precompile. This wires that machinery into the RL trainer as the first step toward compiled + precompiled RL training.

## How

- **`models/simple_fsdp_parallelize.py`** — RL-owned SimpleFSDP `parallelize_fn` (no PP/CP) and `to_simple_fsdp_spec`, which wraps the model config in `GraphTrainerQwen3Model.Config` so weight init runs under `disable_active_parametrization`, while **preserving RL converters** (fp32 `LMHeadCastConverter`, `BatchInvariantFlexConverter`). The spec swap is done inside the trainer, so the generator (separate process) keeps its eager spec.
- **`actors/traced_step.py`** — monarch-free trace/run helper (`RLTracedStep`). Threads GRPO/DAPO's extra per-token tensors (`generator_logprobs`, `advantages`, `loss_mask`) and attention masks as graph inputs, returns `{loss, metrics, grads}` so the controller still gets metrics (graph_trainer discards them), and writes grads into `param.grad` (assign-then-accumulate) so the separate `optim_step` endpoint is unchanged.
- **`ChunkedLossWrapper` -> `ChunkedLossWrapperWithParamGrads`** for the graph path: the base wrapper populates lm_head grads via `param.grad` side effects inside its chunk loop, which do **not** survive graph capture (replay would zero them).
- **`losses/dapo.py`** — divide by `global_valid_tokens` as a **0-d tensor** graph input so `make_fx` does not bake the first step's denominator as a constant (it changes every step). Eager `int`/`None` path unchanged.
- **`precompile.py`** — `save_rl_precompiled` + a single-process CooR driver (`python -m torchtitan.experiments.rl.precompile ...`).
- **Controller** — trainer gets its own `trainer_compile: GraphTrainerCompileConfig` (default disabled), separate from the generator's `compile`.

## Numerical proof (1x H100, `tests/test_precompile_numerics.py`)

Builds the RL qwen3 model under SimpleFSDP and checks, **bitwise**:
- traced `RLTracedStep` == eager ChunkedLoss+GRPO for **loss, all grads, all metrics**;
- base `ChunkedLossWrapper` == `ChunkedLossWrapperWithParamGrads` (the swap is grad-equivalent);
- `global_valid_tokens` is a **live** graph input (doubling it halves the loss, ratio 2.0);
- **CooR save/load roundtrip** reproduces the eager loss + grads bitwise.

```
[OK] loss: bitwise equal
[OK] metric[...] x7: bitwise equal
[OK] grad[...] x74: bitwise equal
[OK] loss(base vs paramgrads): bitwise equal
[OK] doubling gvt halved the loss (ratio=2.00000)
[OK] loss(precompiled vs eager): bitwise equal
ALL CHECKS PASSED
```

The CooR driver was also run standalone (`--flavor debugmodel --attn-backend flex`), producing a loadable artifact.

## Scope / follow-ups

Draft. Validated for **flex attention + dense qwen3 at dp_shard=1**. Not covered here (tracked as follow-ups):
- **varlen** attention: `cu_seqlens` length and `max_q/max_k` (`.item()`) are data-dependent -> needs dynamic-dim marking or fixed-length padding.
- **multi-rank CooR load under Monarch**: each trainer proc must address its GPU as `cuda:0` (torchrun does this via `--virtual-local-rank`); the Monarch provisioner change is not in this PR. The JIT `aot_fx_trace` path needs none of this.
- **dp_shard>1** weight-sync smoke test through SimpleFSDP's spanned mesh.
- **TP loss-parallel** strategy (DTensor `loss_parallel()` vs RL's manual path).
- MoE (gpt_oss / qwen3-MoE) + precompile.

The full RL loop (Monarch + TorchStore + vLLM) was not runnable in the dev env, so validation is via the standalone bitwise test that exercises the exact traced step the actor calls.

## Test plan

```
python -m torchtitan.experiments.rl.tests.test_precompile_numerics
```

### CODE REVIEW

File: torchtitan/experiments/rl/actors/trainer.py [hunk 1]
```diff
@@ -110,8 +115,29 @@ def __init__(
 
         self.config = config
         self.compile_config = compile_config
-        self.loss_fn = config.loss.build()
-        # TODO: add support to compile the loss.
+        # aot_fx_trace mode captures fwd+loss+bwd as one FX graph and requires a
+        # SimpleFSDP-parallelized model. It is opted into by passing a
+        # GraphTrainerCompileConfig with mode="aot_fx_trace"; a plain
+        # CompileConfig (e.g. the generator's shared config) keeps the eager path.
+        self._use_aot_fx_trace = (
```
Comment (felipemello1) [inline torchtitan/experiments/rl/actors/trainer.py:122]: noob question: Does torchtitan/trainer.py (the core trainer) also have this? I couldn't find it. If not, why not? The reason i ask is because i think that eventually we should unify the rl trainer with the core trainer.py

Comment (bobrenjc93) [reply torchtitan/experiments/rl/actors/trainer.py:122]: @felipemello1 graph_trainer does. fwiw this is completely claude written and I haven't reviewed it yet so I wouldn't read too much into this PR yet. I'll take this out of draft mode if/when I think it's ready.
